### PR TITLE
F taint propagation experiment

### PR DESCRIPTION
**Pull request by labrenz**
Allow to run vara's taint analysis on designed set of test files for propagation evaluation.

The compiled cpp files and their filecheck files can be found under [this repo](https://github.com/se-passau/vara-perf-tests/tree/f-taintTests/taint-tests).

The result of each filecheck is stored into a result file for each test file.
If the output of vara matches the expected filecheck an empty file with the `_success` file ending is created, otherwise an `_failed.txt` file will be created. The content of the failed run's output is stored in the result file.

### CODE REVIEW

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- varats/data/reports/taint_report.py  1
- varats/projects/test_projects/taint_tests.py  3
- varats/experiments/taint_propagation.py  4
         

Clones added
============
- varats/projects/c_projects/lrzip.py  1
- varats/experiments/git_blame_annotation_report.py  2
- varats/projects/c_projects/libvpx.py  1
- varats/experiments/taint_propagation.py  2
         

Clones removed
==============
+ varats/projects/c_projects/xz.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/VaRA/VaRA-Tool-Suite/pullRequest?prid=4280681&bid=14603033)

**Comment by labrenz** [comment]
Closing again, while i correctly resolve all merging issues with the vara-dev branch.

**Comment by labrenz** [comment]
Conflicts should all be resolved now.

File: varats/data/reports/taint_report.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+from varats.data.report import BaseReport, FileStatusExtension
+
+
+class TaintPropagationReport(BaseReport):
+    """
+    Print the result of filechecking a llvm ir file generated by VaRA or Phasar
+    in a readable manner.
+    """
+
+    SHORTHAND = "TAINT"
```
Comment (vulder) [inline varats/data/reports/taint_report.py]: maybe we should use something more specific here, TAINTPROP or short TP? or maybe `TPR` ?

File: varats/data/reports/taint_report.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    FILE_TYPE = "txt"
+
+    def __init__(self, path: Path) -> None:
+        super(TaintPropagationReport, self).__init__()
+        with open(path, "r") as stream:
```
Comment (vulder) [inline varats/data/reports/taint_report.py]: why do you open the file as a `stream` but do not use it?

File: varats/data/reports/taint_report.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+                self._path = path
```
Comment (vulder) [inline varats/data/reports/taint_report.py]:
```suggestion
                self.__path = path
```
private vars should have two underscores

File: varats/data/reports/taint_report.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+    @property
+    def path(self) -> Path:
+        """
+        Path to TaintPropagationReport file.
+        """
+        return self._path
```
Comment (vulder) [inline varats/data/reports/taint_report.py]:
```suggestion
        return self.__path
```

File: varats/data/reports/taint_report.py (hunk 1 above)
Comment (vulder) [reply varats/data/reports/taint_report.py]: you used `TPR` later in the `__repr__` impl maybe this is the best solution

File: varats/data/reports/taint_report.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        """
+        Generates a filename for a taint report.
+        """
+        return BaseReport.get_file_name(TaintPropagationReport.SHORTHAND,
+                                        project_name, binary_name,
+                                        project_version, project_uuid,
+                                        extension_type, file_ext)
+
+    def __repr__(self) -> str:
+        return "TPR: " + os.path.basename(self.path)
```
Comment (vulder) [inline varats/data/reports/taint_report.py]:
```suggestion
        return self.SHORTHAND + ": " + os.path.basename(self.path)
```

File: varats/experiments/taint_propagation.py [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+        # Define the output directory.
+        vara_result_folder = self.RESULT_FOLDER_TEMPLATE.format(
+            result_dir=str(CFG["vara"]["outfile"]),
+            project_dir=str(project.name))
+        mkdir("-p", vara_result_folder)
+
+        timeout_duration = '8h'
+
+        for binary_name in project.BIN_NAMES:
+
```
Comment (vulder) [inline varats/experiments/taint_propagation.py]:
```suggestion
```

File: varats/experiments/taint_propagation.py [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+                        binary_name=binary_name,
+                        project_version=str(project.version))))
+
+            if not all_cache_files_present:
+                break
+
+        if not all_cache_files_present:
+            analysis_actions.append(actions.Compile(project))
+            analysis_actions.append(Extract(project))
+            analysis_actions.append(FileCheckExpected(project))
```
Comment (vulder) [inline varats/experiments/taint_propagation.py]: Is it really needed to cache the filecheck outputs? Normally only generating the `.bc` files is expensive, reanalyzing a already generated `.bc` file should be cheap in comparison.

**Review by vulder — CHANGES_REQUESTED** [review]

File: varats/data/reports/taint_report.py (hunk 2 above)
Comment (labrenz) [reply varats/data/reports/taint_report.py]: Resolved in commit 87f44bdb0f23d4317a52197ddc4f6e83ee21949b.

File: varats/data/reports/taint_report.py (hunk 4 above)
Comment (labrenz) [reply varats/data/reports/taint_report.py]: Resolved in commit 87f44bdb0f23d4317a52197ddc4f6e83ee21949b.

File: varats/experiments/taint_propagation.py (hunk 6 above)
Comment (labrenz) [reply varats/experiments/taint_propagation.py]: Resolved in commit 87f44bdb0f23d4317a52197ddc4f6e83ee21949b.

File: varats/experiments/taint_propagation.py (hunk 7 above)
Comment (labrenz) [reply varats/experiments/taint_propagation.py]: Resolved in commit 87f44bdb0f23d4317a52197ddc4f6e83ee21949b.

Removed the txt files from the BC_FILES directory and make use of benchbuilds tmp directory.

**Review by labrenz — COMMENTED** [review]
Reviewed further changes, applied requested feedback and merged the new vara-dev.

File: varats/data/reports/taint_report.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+        """
+        Generates a filename for a taint report.
+        """
+        return BaseReport.get_file_name(TaintPropagationReport.SHORTHAND,
+                                        project_name, binary_name,
+                                        project_version, project_uuid,
+                                        extension_type, file_ext)
+
+    def __repr__(self) -> str:
+        return self.SHORTHAND + ":" + os.path.basename(self.path)
```
Comment (vulder) [inline varats/data/reports/taint_report.py]:
```suggestion
        return self.SHORTHAND + ": " + os.path.basename(self.path)
```
I normally find a space after a colon easier to comprehend and copy. If you want it without or if the formatting is necessary, it's fine with me.

File: varats/experiments/vara_full_mtfa.py [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+
+
+class VaRATaintPropagation(VaRAVersionExperiment):
+    """
+    Generates a taint flow analysis (MTFA) of the project(s) specified in the
+    call.
+    """
+
+    NAME = "VaRATaintPropagation"
+
```
Comment (vulder) [inline varats/experiments/vara_full_mtfa.py]:
```suggestion
```

**Review by vulder — APPROVED** [review]
Please fix the two nits otherwise it's fine, I'll merge it afterwards.

**Pull request merged**